// @laufire/utils@3.0.0-0/random.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stringSeeds = exports.rndValueWeighted = exports.rndValues = exports.rndValue = exports.rndOfString = exports.rndString = exports.rndBetween = void 0;

var _collection = require("./collection");

var _lib = require("./lib");

var _fn = require("./fn");

/**
 * Random functions to help with testing.
 */

/* Imports */

/* Exports */
var stringSeeds = {
  "char": 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
  hex: '0123456789ABCDEF',
  num: '0123456789'
};
/**
 *
 * @param {Integer} [0] from - The start of the range.
 * @param {Integer} [1] to - The end of the range.
 * @returns {integer} A random number between from and to.
 */

exports.stringSeeds = stringSeeds;

var rndBetween = function rndBetween() {
  var from = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var to = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 10;
  var precision = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  // eslint-disable-next-line no-magic-numbers
  var magnitude = Math.pow(10, precision);
  var result = (0, _lib.rndBetween)(from * magnitude, to * magnitude) / magnitude;
  return Number(result.toFixed(precision));
};
/*
	NOTE: To by-pass seed matching of predefined names, use them twice.
	IE: 'charchar' instead of 'char'.
*/

/**
 * Get a random string from the pre-defined seed / custom string.
 * @param {integer} [8] length
 * @param {string} [char] seed
 * @returns { string} A random string.
 */
// eslint-disable-next-line no-magic-numbers


exports.rndBetween = rndBetween;

var rndString = function rndString() {
  var length = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 8;
  var seed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'char';
  var seedString = stringSeeds[seed] || seed;
  var seedCharCount = seedString.length;
  var ret = ''; // eslint-disable-next-line no-param-reassign

  while (length--) {
    ret += seedString.substr(rndBetween(0, seedCharCount), 1);
  }

  return ret;
};

exports.rndString = rndString;

var rndValue = function rndValue(collection) {
  var items = (0, _collection.values)(collection);
  return items[rndBetween(0, items.length)];
};

exports.rndValue = rndValue;

var rndValues = function () {
  var skipRndKey = function skipRndKey(iterable) {
    var rndKey = rndValue((0, _collection.keys)(iterable));
    return (0, _collection.filter)(iterable, function (dummy, key) {
      return key !== rndKey;
    });
  };

  return function (iterable, count) {
    var length = (0, _fn.defined)(count, rndBetween(0, iterable.length + 1));
    return (0, _collection.reduce)(iterable, function (t) {
      return (0, _collection.keys)(t).length > length ? skipRndKey(t) : t;
    }, iterable);
  };
}();
/**
 * Get a random substring from the given string.
 * @param {string} string - The seed string.
 * @returns {string} The resulting random substring.
 */


exports.rndValues = rndValues;

var rndOfString = function rndOfString(string, length) {
  return (0, _collection.values)(rndValues(string.split(''), length)).join('');
};

exports.rndOfString = rndOfString;

var rndValueWeighted = function rndValueWeighted(weights) {
  var candidates = (0, _collection.keys)(weights);
  var boundaries = (0, _collection.values)((0, _collection.map)(weights, function (value, key) {
    return value + candidates.slice(0, candidates.indexOf(key)).reduce(function (t, c) {
      return t + weights[c];
    }, 0);
  }));
  var start = 0;
  var end = boundaries.slice(-1)[0] || 1;
  return function () {
    var marker = rndBetween(start, end);
    return candidates[boundaries.findIndex(function (boundary) {
      return boundary > marker;
    })];
  };
};

exports.rndValueWeighted = rndValueWeighted;
