// @laufire/utils@3.0.0-0/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isArray = exports.rndBetween = exports.values = exports.keys = exports.entries = exports.assign = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var assign = Object.assign,
    objEntries = Object.entries,
    values = Object.values,
    objKeys = Object.keys;
exports.values = values;
exports.assign = assign;
var floor = Math.floor,
    random = Math.random;
var ownKeys = Reflect.ownKeys;
var isArray = Array.isArray;
exports.isArray = isArray;

var keys = function keys(collection) {
  return isArray(collection) ? objKeys(collection).map(Number) : ownKeys(collection);
};

exports.keys = keys;

var entries = function entries(collection) {
  return isArray(collection) ? objEntries(collection).map(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
        key = _ref2[0],
        value = _ref2[1];

    return [Number(key), value];
  }) : objEntries(collection);
}; // eslint-disable-next-line no-magic-numbers


exports.entries = entries;

var rndBetween = function rndBetween(from, to) {
  return floor(random() * (to - from)) + from;
};

exports.rndBetween = rndBetween;
